### ISSUE TITLE
Next chunks (SIM jobs) getting submitted before finishing the previous chunk

### ISSUE THREAD

**Issue by kinow** [BSC-ES/autosubmit#642]
In GitLab by @ rmahmood on Jan 9, 2021, 18:25

Hi @mcastril , @pabretonniere , @wuruchi  

I am trying to extend the number of chunks from 2 to 10 in the experiment a330. First two chunks were run initially for all start dates and now I want to extend the experiment to a total of 10 chunks (only for 5 start-dates).  

I had suspended `CLEAN_MEMBER` and `TRANSFER_MEMBER` so that the runtime contains files from previously finished chunk. 

I am wondering why autosubmit is submitting not only the third chunk, which is running, but also submitting chunk 4 and 5 which would fail because chunk 3 is not finished. 

In expdef_a330.conf I am still using 'RERUN=FALSE', because it is not about rerunning a previous chunk.

I did look at FAQ (https://earth.bsc.es/gitlab/es/auto-ecearth3/-/wikis/extend_experiment), however I guess this does not apply in my case because the runtime files are fine and also the workflow in ece.info is fine (and the chunk 3 seems to be running fine). Or maybe I am missing something.

`http://bscesweb04.bsc.es/autosubmitapp/experiment/a330`

Thanks,

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 11, 2021, 11:02

Hi @ rmahmood , can you print the Autosubmit log of the run? So we can see which was the status before that chunks were submitted and run (it should be WAITING, not READY).

Please do not use RERUN, it lacks support right now.

You can tag again @dbeltrankyl now that he's back.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 11, 2021, 11:25

Hello @ rmahmood @mcastril 

I can see why `SIM 3` was submitted at the same time as `SIM 4`.

Since `CLEAN 2` was `COMPLETED`, and because of its dependencies, then that implies that `SIM 2` and `SIM 3` are `COMPLETED`. Also, since `CLEAN 1` was `COMPLETED`, then `SIM 4` can start at the same time as `SIM 3` even though `SIM 3` was not completed. I guess this is what happened to `SIM 4`, and I suspect something similar happened to `SIM 5`.

In short, due to the status of the already ran jobs and the increase in chunks, some dependencies might have been ignored due to the constraints satisfied by the status of already executed jobs.

**Comment by kinow** [comment]
In GitLab by @ rmahmood on Jan 11, 2021, 11:53

Hi @mcastril , @wuruchi , Thanks!

Here are the logs which shows indeed the chunk 4 and 5 were set to READY by the `recovery`:

> a330_19601101_fc00_1_SIM COMPLETED /tmp/LOG_/a330/ /tmp/LOG_/a330/
a330_19601101_fc00_2_SIM COMPLETED /tmp/LOG_/a330/ /tmp/LOG_/a330/
a330_19601101_fc00_3_SIM READY  
a330_19601101_fc00_4_SIM READY  
a330_19601101_fc00_5_SIM READY  
a330_19601101_fc00_6_SIM WAITING  
a330_19601101_fc00_7_SIM WAITING  
a330_19601101_fc00_8_SIM WAITING  
a330_19601101_fc00_9_SIM WAITING  
a330_19601101_fc00_10_SIM WAITING

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 11, 2021, 12:12

> Since `CLEAN 2` was `COMPLETED`, and because of its dependencies, then that implies that `SIM 2` and `SIM 3` are `COMPLETED`.

I think that if SIM 2 is the last chunk, CLEAN 2 can run even if it indirectly depends on SIM+1. If not, experiments would not have their last chunk cleaned.

> Also, since `CLEAN 1` was `COMPLETED`, then `SIM 4` can start at the same time as `SIM 3` even though `SIM 3` was not completed.

Are you sure? There is a dependency on a previous SIM: `SIM = SIM -1`. So in case that SIM 3 is not COMPLETED, SIM 4 should have to wait.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 11, 2021, 12:14

Sorry if this is obvious but I don't have rememberings about this now... I have a general doubt about https://earth.bsc.es/gitlab/es/auto-ecearth3/-/wikis/extend_experiment now:

If you are to extend the last chunk, shouldn't you re-run the POST of the last chunk? To include the first step of the next chunk in the result.

@pabretonniere @ aamaral @etiennesky

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 11, 2021, 12:22

Same here:

```
2021-01-09 15:47:41,984 CHANGED job 'a330_19601101_fc00_1_SIM' status to COMPLETED
2021-01-09 15:47:42,766 CHANGED job 'a330_19601101_fc00_2_SIM' status to COMPLETED
2021-01-09 15:47:43,560 CHANGED job 'a330_19601101_fc00_3_SIM' status to WAITING
2021-01-09 15:47:44,295 CHANGED job 'a330_19601101_fc00_4_SIM' status to WAITING
2021-01-09 15:47:45,042 CHANGED job 'a330_19601101_fc00_5_SIM' status to WAITING
.
.
.
2021-01-09 16:17:19,016 Resetting job: a330_19601101_fc00_3_SIM status to: READY (all parents completed)...
2021-01-09 16:17:19,016 Resetting job: a330_19601101_fc00_4_SIM status to: READY (all parents completed)...
2021-01-09 16:17:19,016 Resetting job: a330_19601101_fc00_5_SIM status to: READY (all parents completed)...
```

`/esarchive/autosubmit/a330/tmp/20210109_154657_recovery.log`


I don't see how this can be compatible to:

```
[SIM]
FILE = sources/runtime/autosubmit/ece-esm.sh
DEPENDENCIES = INI SIM-1 CLEAN-3 NCTIME-3
RUNNING = chunk
WALLCLOCK = 2:00
PROCESSORS = 768
```

Any ideas @dbeltrankyl @wuruchi ?

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 11, 2021, 12:40

> I think that if SIM 2 is the last chunk, CLEAN 2 can run even if it indirectly depends on SIM+1. If not, experiments would not have their last chunk cleaned.

Yeah, I am sure `CLEAN 2` was completed in the previous run, but then the user added more chunks so there was a `SIM 3`. In consequence, logically that meant (for the dependency tree) that `SIM 3` had been completed (although it was not) and then `SIM 4` could start. My point is that by adding more chunks, some constraints were satisfied that allowed `SIM 4` and `SIM 5` to run, which under normal sequential conditions (starting the run from the beginning with all the chunks considered) should have waited.

> Are you sure? There is a dependency on a previous SIM: `SIM = SIM -1`. So in case that SIM 3 is not COMPLETED, SIM 4 should have to wait.

I see that `SIM 4` depended on `SIM 3`, but `CMORATM 2` depended on `SIM 3`, and `CLEAN 1` depended on `CMORATM 2`. Thus, by transitive reduction, this dependency tree is simplified to `SIM 4` depends on `CLEAN 1`, which is correct. However, since in the previous run `CLEAN 1` was completed because there were fewer chunks, with the added chunks, `SIM 4` was allowed to start.

Here is the graph without transitive reduction:

![image](https://earth.bsc.es/gitlab/es/autosubmit/uploads/0b64a87d7df1fbb99d1c0625e03d6c88/image.png)

Now, with the **no transitive reduction** graph, we can track why `SIM 5` started:

* `SIM 5` depends on `SIM 4`
* `CMORATM 3` depends on `SIM 3` and `SIM 4`
* `CLEAN 2` depends on `CMORATM 3`
* `SIM 5` depends con `CLEAN 2`

By transitive reduction, this is simplified to `SIM 5` depends on `CLEAN 2`. Nonetheless, `CLEAN 2` was already completed in the run with **fewer chunks**, so `SIM 5` was allowed to start.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 11, 2021, 16:24

I see @wuruchi , I understood it once I read the first paragraph now. As you were not mentioning the transitive reduction nor the dependency tree in the morning (I was also confusing to read that CLEAN 3 was COMPLETED when it wasn't), I was not realising that this is the only thing Autosubmit takes to generate the new dependencies, but you were right from the beginning. 

So this is something intrinsic to Autosubmit because we cannot make it look behind the pruned dependencies unless we set -no-transitive right? But this would be too much for such a big experiment.

Our recommendation is always to look the workflow situation after a recovery or a chunk extension is performed, because in a different workflow things may change as we saw know. @ rmahmood you should set the last (2nd chunk) CLEAN and maybe POST to waiting before relaunching so SIM_4, etc are not put to RUN.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 11, 2021, 18:43

I've been thinking a little bit more about this. Do you think that you could add a warning by console output whenever a user does a RECOVERY and there are jobs `WAITING` which some of their children are COMPLETED? 

I think this would be beneficial for a lot of cases, and maybe not so much computationally expensive. This situation can appear on purpose whenever a user wants to RERUN a particular chunk but in that case, I think setstatus is handier. I think that adding such warning in recovery will bring more benefits that drawbacks.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 12, 2021, 10:31

>So this is something intrinsic to Autosubmit because we cannot make it look behind the pruned dependencies unless we set -no-transitive right? But this would be too much for such a big experiment.

Yes, no-transitive is required to check all previous dependencies.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 12, 2021, 10:34

Hello @mcastril 

> I've been thinking a little bit more about this. Do you think that you could add a warning by console output whenever a user does a RECOVERY and there are jobs `WAITING` which some of their children are COMPLETED?

Implementing this warning is possible, although it might be restricted by the result of the no-transitive reduction. We can add it to the pending list.

**Comment by kinow** [comment]
In GitLab by @mcastril on Jan 12, 2021, 10:42

Thanks @wuruchi .

Yes, given that we are constrained by the reduced graph, we can at least tell when there is a not COMPLETED parent tree with some COMPLETED children, which is an abnormal situation wrt to the usual workflow status. If it's on purpose then the user can just ignore it, but we can save data from being removed, runtime directories from being modified, etc. 

And recovery is most of the times used to recover a workflow which has been running normally, not as setstatus that is more to modify or alter a workflow in any way possible, for re-runs and so on, so keeping the warning in recovery I think it won't produce much disturbance.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 15, 2021, 12:42

mentioned in commit f07aa51cc0596429b7686f27f5c6313f0719e3d5

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 15, 2021, 12:49

A warning has been implemented (`3.13.0b`) when conflicting status on dependencies arises after recovery is executed.

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 15, 2021, 23:55

closed via merge request !233

**Comment by kinow** [comment]
In GitLab by @wuruchi on Jan 15, 2021, 23:55

mentioned in commit 115bf2f80f3755f390650d51ed7574123c1559cd
